### ISSUE TITLE
zdb -h | -i crashes (seg fault)

### ISSUE THREAD

**Issue by don-brady** [zfsonlinux/zfs#3520]
This is a regression from the recent large block support commit f1512ee6 (illumos issue 5027).

This crash is very easy to reproduce for "-h" case.  The "-i" crash requires an active ZIL log with a dirty `TX_WRITE` record. The ziltest tool easily provides this condition and will crash when it invokes  zdb -ivv  to display the transactions to replay.  Both crash cases are shown below.

Currently testing a fix and will submit a pull request soon.

```
//
// case '-h' crashes in dump_history()
//
[root]# zdb -h mercury
Segmentation fault (core dumped)
```

```
//
// case '-i' crashes in zil_prt_rec_write()
//
[root]# zdb -ep /tmp -ivv ziltestpool.48157/fs
    ZIL header: claim_txg 0, claim_blk_seq 0, claim_lr_seq 0 replay_seq 0, flags 0x0

        TX_CREATE           len    112, txg 12, seq 1
            a
            Wed Jun 24 10:21:37 2015
            doid 4, foid 9, mode 100644
            uid 0, gid 0, gen 12, rdev 0x0
        TX_SETATTR          len    112, txg 12, seq 2
            foid 9, mask 0x300
            AT_ATIME 1435162897.709260159 Wed Jun 24 10:21:37 2015
            AT_MTIME 1435162897.709260159 Wed Jun 24 10:21:37 2015
<snip>
        TX_WRITE            len    272, txg 12, seq 36
Segmentation fault (core dumped)
```

**Comment by behlendorf** [comment]
The root might be the same but @lundman also reported a issue with `zdb` which was introduced by large blocks in #3468.  Since you're already working in this bit of the code it would be great if you could take a look!

**Comment by behlendorf** [comment]
Resolved by 017da6f063861a1dc3eea5a41c3de8b879d579bd.
